pandas/indexes/numeric.py

```diff
@@ -154,14 +154,86 @@ def _assert_safe_casting(cls, data, subarr):
         """
         Ensure incoming data can be represented as ints.
         """
-        if not issubclass(data.dtype.type, np.integer):
+        if not issubclass(data.dtype.type, np.signedinteger):
             if not np.array_equal(data, subarr):
                 raise TypeError('Unsafe NumPy casting, you must '
                                 'explicitly cast')
 
 Int64Index._add_numeric_methods()
 Int64Index._add_logical_methods()
 
+_uint64_descr_args = dict(
+    klass='UInt64Index',
+    ltype='unsigned integer',
+    dtype='uint64',
+    extra=''
+)
+
+
+class UInt64Index(NumericIndex):
+    __doc__ = _num_index_shared_docs['class_descr'] % _uint64_descr_args
+
+    _typ = 'uint64index'
+    _arrmap = _algos.arrmap_uint64
+    _left_indexer_unique = _join.left_join_indexer_unique_uint64
+    _left_indexer = _join.left_join_indexer_uint64
+    _inner_indexer = _join.inner_join_indexer_uint64
+    _outer_indexer = _join.outer_join_indexer_uint64
+
+    _can_hold_na = False
+    _na_value = 0
+
+    _engine_type = _index.UInt64Engine
+
+    _default_dtype = np.uint64
+
+    @property
+    def inferred_type(self):
+        return 'integer'
+
+    @property
+    def asi8(self):
+        # do not cache or you'll create a memory leak
+        return self.values.view('u8')
+
+    @Appender(_index_shared_docs['_convert_scalar_indexer'])
+    def _convert_scalar_indexer(self, key, kind=None):
+        assert kind in ['ix', 'loc', 'getitem', 'iloc', None]
+
+        # don't coerce ilocs to integers
+        if kind != 'iloc':
+            key = self._maybe_cast_indexer(key)
+        return (super(UInt64Index, self)
+                ._convert_scalar_indexer(key, kind=kind))
+
+    @Appender(_index_shared_docs['_convert_arr_indexer'])
+    def _convert_arr_indexer(self, keyarr):
+        if is_integer_dtype(keyarr):
+            return _asarray_tuplesafe(keyarr, dtype=np.uint64)
+        return keyarr
+
+    @Appender(_index_shared_docs['_convert_index_indexer'])
+    def _convert_index_indexer(self, keyarr):
+        if keyarr.is_integer():
+            return keyarr.astype(np.uint64)
+        return keyarr
+
+    def _wrap_joined_index(self, joined, other):
+        name = self.name if self.name == other.name else None
```

Comment:
this should prob be handled more generically in the super method (e.g. why are we not just calling ``_simple_new(joined...)``?